pkg/controller/registryinfra/registryinfra_controller.go

```diff
@@ -0,0 +1,696 @@
+		})
+		if err != nil {
+			return fmt.Errorf("error listing user inline policies: %v", err)
+		}
+		for _, policy := range userPolicies.PolicyNames {
+			_, err = iamSvc.DeleteUserPolicy(&iam.DeleteUserPolicyInput{
+				UserName:   aws.String(userName),
+				PolicyName: policy,
+			})
+			if err != nil {
+				return fmt.Errorf("error deleting inline user policies: %v", err)
+			}
+		}
+
+		return nil
+	}
+}
+
+func createS3IamPolicy(bucketName string) (string, error) {
+	t, err := template.New("s3policy").Parse(s3PolicyTemplate)
+	if err != nil {
+		return "", fmt.Errorf("error parsing s3policy template: %v", err)
+	}
+
+	params := s3PolicyTemplateParams{
+		BucketName: bucketName,
+	}
+
+	var buf bytes.Buffer
+	err = t.Execute(&buf, params)
+	if err != nil {
+		return "", fmt.Errorf("error processing s3policy template: %v", err)
+	}
+	policyDocument := buf.String()
+
+	return policyDocument, nil
+}
+
+func (c *Controller) setupRegistryInfra(cluster *clusteroperator.CombinedCluster, createInfra bool) (*iam.AccessKey, error) {
+	clusterID := cluster.Name
+
+	// name for secret where we will store the registry object store credentials
+	objectStoreSecretName := controller.RegistryCredsSecretName(clusterID)
+
+	registrySecret, err := c.kubeClient.CoreV1().Secrets(cluster.Namespace).Get(objectStoreSecretName, metav1.GetOptions{})
+	secretFound := (err == nil)
+	unexpectedError := (err != nil && !errors.IsNotFound(err))
+
+	if unexpectedError {
+		return nil, fmt.Errorf("error querying for existing registry secret: %v", err)
+	} else if secretFound && createInfra {
+		storageKey := string(registrySecret.Data["awsAccessKeyId"])
+		storageSecret := string(registrySecret.Data["awsSecretAccessKey"])
+		return &iam.AccessKey{
+			AccessKeyId:     &storageKey,
+			SecretAccessKey: &storageSecret,
+		}, nil
+	}
+
+	adminCredsSecretName := cluster.AWSClusterProviderConfig.Hardware.AccountSecret.Name
+	bucketName := controller.RegistryObjectStoreName(clusterID)
+
+	secret, err := c.kubeClient.CoreV1().Secrets(cluster.Namespace).Get(adminCredsSecretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving aws creds secret: %v", err)
+	}
+	accessKey := string(secret.Data["awsAccessKeyId"])
+	secretAccessKey := string(secret.Data["awsSecretAccessKey"])
+
+	awsCreds := credentials.NewStaticCredentials(accessKey, secretAccessKey, "")
+	config := aws.Config{Region: &cluster.AWSClusterProviderConfig.Hardware.Region,
+		Credentials: awsCreds}
+	sess := session.New(&config)
+
+	// create/delete bucket if necessary
+	err = c.setupS3Bucket(bucketName, sess, createInfra)
+	if err != nil {
+		return nil, err
+	}
+
+	// create/delete user if necessary
+	newAccessKey, err := c.setupS3IamUser(clusterID, bucketName, sess, createInfra)
+	if err != nil {
+		return nil, err
+	}
+
+	// save/delete iam user's AWS creds into secret
+	err = c.setupRegistrySecret(objectStoreSecretName, cluster.Namespace, newAccessKey, createInfra)
```

Comment:
if we failed on secret creation, we'd create yet another aws_key/aws_secret_key pair since you only get one opportunity to receive the aws_secret_key. there is a limit on the number of credentials you can generate for a user... maybe delete any existing keys before we generate??